### ISSUE TITLE
Add config to let users specify key_length, charset, and actions when links are unshortened.

### ISSUE THREAD

**Issue by msquitieri** [jpmcgrath/shortener#25]
Hello,

I think this is a great gem but would love to specify the key_length to however long I'd like as well as the charset to increase the number of total possible combinations. It might be best to change the defaults to a key length of 6 and a case sensitive charset so that the total combinations will be over 50 billion, which will cover pretty much any use case.

Also, in addition to saving the number of times the link was unshortened it would be great for users to be able to provide their own callback event to log other information or perform other actions when the link is unshortened.

Thanks!

**Comment by jpmcgrath** [comment]
@msquitieri just following up on these suggestions.

WRT specifying the dafault key length, does the existing functionality of setting Shortener.unique_key_length in your initializer do the trick? If not, are/were you after the ability to set a key length wheneever you generated the URL?

When you wrote "great for users to be able to provide their own callback event", when you write "users", are you talking about developers using the Shortener Gem? 

Thanks.

**Comment by jpmcgrath** [comment]
For the callback, I am thinking along these lines:

Shortener::ShortenedLink.generate('http://example.com', callback: 'puts "hello world!"')

The callback string would be stored in the shortened link record. Then in the ShortenedLink controller show action we would call ```eval(shortened_link.callback)```

That said, this is pretty dangerous code, it will be important to ensure end users cannot set the value of the callback value, otherwise they could wreak havok with a system. Perhaps this functionality is off by default.

**Comment by jpmcgrath** [comment]
Also, on the point about changing the charset, there has been the ability for quite a while (although not mentioned in the readme), to set the charset to be alphanumeric case insensitve. I have updated the readme to reflect this.

**Comment by msquitieri** [comment]
Let me go down the list:

- Yes, setting `Shortener.unique_key_length` in your initializer does do the trick. I think I discovered that after opening up this issue.

- When I wrote "users" I meant developers who use the gem.

- In terms of the callback, I though it could be set in an initializer:

```ruby
# initializers/shortener.rb

Shortener.configure do |c|
  c.after_unpack do |shortened_link|
    # do something with the link.
    # like log to your own source
  end

  c.before_unpack do |shortened_link|
    # do something
  end

end
```

I have tried doing something like above on a branch of the gem, but I had issues with the scope of the block that is passed, where some application code was not defined in the scope. This is something to consider however.

**Comment by jpmcgrath** [comment]
Right, I understand now. I'll have a fiddle over the next few days to see what I can come up with.

**Comment by jpmcgrath** [comment]
What do you think of this approach:

1. Modify the Shortener::ShortenedUrlsController show action to return the shortened link record.
2. Create a controller that subclasses Shortener::ShortenedUrlsController and overrides the show action:

  class CustomShortenerController < Shortener::ShortenedUrlsController
    def show
      # do something
      short_link = super
      # do something with the returned sort link
      # like write a log entry
    end
  end

3. Make the shortened link routes point to this new controller action:

   get '/:id' => "custom_shortener#show"

This would allow you to write log messages, or do whatever you like when the link is unshortened. It would also provide a lot of flexibility for the Gem user, without having to modify the behaviour of the gem very much.

**Comment by msquitieri** [comment]
Hmm, I'm not sure you would want the `Shortener::ShortenedUrlsController#show` action to return the record. Controller actions don't usually return values. Also that action also tells the request to redirect to the unpacked URL, which, if you're customizing it, you might not want to happen. We want to keep the plug-and-play case where someone just drops in the gem and expects it to work.

Looking at the controller action, I see a couple of things happening:
- fetch the url
- increment the count
- merge query params
- redirect to new url

In the customization case, fetching the URL and redirecting the URL can be left to the implementer. However, incrementing and setting the correct query params should be part of the model. That's why I recommend moving those into their own methods on the model. Maybe something like `Shortened::ShortenedUrl.fetch_url_with_token(token, params={})` which would accept the unique_key and a set of query params. This would find the object with that token and return the decorated url with the correct query params that were passed.

Adding this, and explaining it in the README how to customize it should be enough.

**Comment by jpmcgrath** [comment]
Good ideas. I'll have a crack at implementing today.

**Comment by jpmcgrath** [comment]
@msquitieri  https://github.com/jpmcgrath/shortener/pull/51 Comments?

Note, this feature build upon some refactoring I did in a prior PR. The code changes related to this discussion are best seen here: https://github.com/jpmcgrath/shortener/compare/bugfix/custom-key-clobbering...feature/show-action-on-diet

**Comment by jpmcgrath** [comment]
published in v0.5.5.

**Comment by msquitieri** [comment]
Sorry for the late, response. This looks good! My only comment would be to make `::Shortener::ShortenedUrl.fetch_with_token(token: token, additional_params: params)` take the `token` as a required parameter, since it does not make much sense for that to be passed in an options hash.
